### PR TITLE
Basic CHC Encoding

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — CHANGES_REQUESTED** [review]
1) I'm fine with the first two commits (those adding new constructors).
2) I find "substitute_symbols" very confusing as we already have "replace_symbolt". Merriam-Webster says "replace" and "substitute" are synonyms.
3) I don't have the necessary background to comment on the CHC encoder, but not having any tests does not seem appropriate.

**Comment by kroening** [comment]
2. The key difference is that the subsitute_symbols does consider bindings. This is needed in most cases, but too much of a change to do to replace_symbols.

**Comment by kroening** [comment]
3. Now with tests.

**Comment by tautschnig** [comment]
> 3. Now with tests.

Thank you!

**Comment by tautschnig** [comment]
> 2. The key difference is that the subsitute_symbols does consider bindings. This is needed in most cases, but too much of a change to do to replace_symbols.

It may well be that having two separate implementations is the right thing to have, but then their names need to be sufficiently clear to avoid any confusion.

Also, I would expect the new code to handle all `binding_exprt` (it's missing `array_comprehension_exprt` - with #6810 one could just use `if(can_cast_expr<binding_exprt>(...)` or `expr_try_dynamic_cast<binding_exprt>` to avoid this problem), and the existing code in `std_expr.cpp` to be removed.

**Comment by kroening** [comment]
'substitute' is the common term for replacing a variable by an expression; 'replace' is the unusual one. We will likely gradually replace (uh) 'replace_symbol' by 'substitute_symbol', since we need that awareness of bindings.

https://en.wikipedia.org/wiki/Substitution_(logic)

I've added the array_comprehension case.

**Comment by tautschnig** [comment]
> 'substitute' is the common term for replacing a variable by an expression; 'replace' is the unusual one. We will likely gradually replace (uh) 'replace_symbol' by 'substitute_symbol', since we need that awareness of bindings.
> 
> https://en.wikipedia.org/wiki/Substitution_(logic)
> 

But then the current state of affairs is that we have `address_of_exprt`, which `replace_symbolt` gets right, but `substitute_symbol` does not. At the same time, `replace_symbolt` does not know about bound symbols when `substitute_symbol` does. We, therefore, have two implementations the names of which are synonyms, their features are orthogonal, and neither of them really works on the current code base.

> I've added the array_comprehension case.

Thank you!

**Comment by kroening** [comment]
Right, we'll need to manage the transition.
How about adding binding-awareness to `replace_symbol(...)`?

**Comment by tautschnig** [comment]
> Right, we'll need to manage the transition. How about adding binding-awareness to `replace_symbol(...)`?

This would help avoid confusion, and then we could still introduce a (thin) wrapper `substitute_symbol` with the API proposed in this PR that just uses `replace_symbol` under the hood? That wrapper would, among other things, help fix up terminology.

**Comment by kroening** [comment]
No, we want to get rid of `replace_symbol` in the longer term -- so that should be a wrapper around `substitute_symbol`.

**Comment by kroening** [comment]
@tautschnig -- shall we go for that?

**Comment by kroening** [comment]
Now without the offending commit.

File: src/goto-instrument/horn_encoding.cpp [hunk 1]
```diff
@@ -13,10 +13,1203 @@ Date: June 2015
...
+
+#include <goto-programs/goto_model.h>
+
+#include <solvers/smt2/smt2_conv.h>
+
+#include <algorithm>
 #include <ostream>
+#include <unordered_set>
+
+class state_typet : public typet
```
Comment (peterschrammel) [inline src/goto-instrument/horn_encoding.cpp:34]: 📖 A bit of documentation would be helpful for the next reader.

**Review by peterschrammel — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]

**Pull request merged**